### PR TITLE
refactor(api): protocol_api package shuffling

### LINKED ISSUE #6331: refactor(api): separate public facing protocol_api classes from implementation

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

This PR takes everything out of `protocol_api` package that isn't a context object. 

The rationale is that CPX will be rebuilding the context objects. Everything in `protocol_api` will be about docstrings and input validation. The implementation will reside elsewhere. 

# Changelog

- execution related modules moved to `protocols.execution`
- api support (ie geometry, transfers, utils) moved to `protocols.api_support`

Original structure
```
opentrons
├── protocol_api
│   ├── __init__.py
│   ├── constants.py
│   ├── contexts.py
│   ├── definitions.py
│   ├── dev_types.py
│   ├── execute.py
│   ├── execute_v3.py
│   ├── execute_v4.py
│   ├── execute_v5.py
│   ├── geometry.py
│   ├── instrument_context.py
│   ├── json_dispatchers.py
│   ├── labware.py
│   ├── module_contexts.py
│   ├── module_geometry.py
│   ├── protocol_context.py
│   ├── transfers.py
│   ├── types.py
│   └── util.py
```

New structure

```
opentrons
├── protocol_api
│   ├── __init__.py
│   ├── contexts.py
│   ├── instrument_context.py
│   ├── labware.py
│   ├── module_contexts.py
│   ├── paired_instrument_context.py
│   └── protocol_context.py
├── protocols
│   ├── __init__.py
│   ├── advanced_control
│   │   ├── __init__.py
│   │   └── transfers.py
│   ├── api_support
│   │   ├── __init__.py
│   │   ├── constants.py
│   │   ├── definitions.py
│   │   └── util.py
│   ├── bundle.py
│   ├── execution
│   │   ├── __init__.py
│   │   ├── dev_types.py
│   │   ├── errors.py
│   │   ├── execute.py
│   │   ├── execute_json_v3.py
│   │   ├── execute_json_v4.py
│   │   ├── execute_json_v5.py
│   │   ├── execute_python.py
│   │   ├── json_dispatchers.py
│   │   └── types.py
│   ├── geometry
│   │   ├── __init__.py
│   │   ├── deck.py
│   │   ├── deck_item.py
│   │   ├── module_geometry.py
│   │   └── planning.py
│   ├── parse.py
│   ├── types.py
│   └── util.py
```

# Review requests

How do we feel about new package locations?
Is there a reason we shouldn't be doing this?

# Risk assessment

should be none since `ProtocolContext` is the only object py protocol's interact with.

closes #6331

### CODE REVIEW

**Review by Laura-Danielle — COMMENTED** [review]
You can ignore this suggestion, but maybe rather than throwing everything into `api_support`, we can make paths like:

`protocols.geometry` -> holds geometry stuff
`protocols.advanced_control`  -> holds transfers and other non-basic protocol commands we might be adding later
`protocols.api_support` -> Can use this to hold definitions, types etc

Otherwise, the shuffling looks good.

**Comment by sfoster1** [comment]
> You can ignore this suggestion, but maybe rather than throwing everything into `api_support`, we can make paths like:
> 
> `protocols.geometry` -> holds geometry stuff
> `protocols.advanced_control` -> holds transfers and other non-basic protocol commands we might be adding later
> `protocols.api_support` -> Can use this to hold definitions, types etc
> 
> Otherwise, the shuffling looks good.

Yeah I think there's some interesting interactions we could unpack here.

APIv2 is at its core organized into two layers:
- The "hardware control" layer exposes domain-specific actions in a language of robotics in a composable way. Its motion parameters, for functions that need them, are vectors in deck coordinates and specific hardware features (i.e. which mount to control). It organizes, essentially, the gantry and everything attached to it, and modules by their core function.
- the "protocol" layer exposes domain-specific actions in a language of biology. Its two biggest requirements are to expose the verbs of the OT-2 system (like "aspirate", "dispense") at a level relevant to biology, and to provide the nouns like "well A1 of the labware in this well" that allow those verbs to raise to that higher level.

A big part of the refactors we're engaging in is to split the protocol layer into multiple side-by-side implementations with different levels of input checking, but only really within that layer. We may add more verbs (like a better first-class treatment of "move to") but the protocol layer as a whole retains its job of essentially translating the high-level language of biology into the medium-level language that the hardware controller speaks. 

That means that something like `labware.py` and the rest of the geometry system are, well, more than just utilities. They're in fact fundamental to the split between the layers because they provide the nouns. 

This is maybe a lot of words to say that it doesn't make sense to me to move `geometry` and `module_geometry` to the `protocols` package  but keep `labware` in the `protocol_api` area. And it may in fact make the most sense to carve out a different package for the geometry section so that under `opentrons.protocols` we have
- A subpackage for the nouns of the protocol layer (note: not protocol api)
- A subpackage for the interfaces to the verbs of the protocol layer
   - perhaps with further subpackages for implementations, or they're somewhere else
   - maybe this is where the protocol_api definitions live and we preserve `opentrons.protocol_api` as a dummy with an `__init__` that doesn't do much other than import the lower level packages

**Comment by amitlissack** [comment]
> > You can ignore this suggestion, but maybe rather than throwing everything into `api_support`, we can make paths like:
> > `protocols.geometry` -> holds geometry stuff
> > `protocols.advanced_control` -> holds transfers and other non-basic protocol commands we might be adding later
> > `protocols.api_support` -> Can use this to hold definitions, types etc
> > Otherwise, the shuffling looks good.
> 
> Yeah I think there's some interesting interactions we could unpack here.
> 
> APIv2 is at its core organized into two layers:
> 
> * The "hardware control" layer exposes domain-specific actions in a language of robotics in a composable way. Its motion parameters, for functions that need them, are vectors in deck coordinates and specific hardware features (i.e. which mount to control). It organizes, essentially, the gantry and everything attached to it, and modules by their core function.
> * the "protocol" layer exposes domain-specific actions in a language of biology. Its two biggest requirements are to expose the verbs of the OT-2 system (like "aspirate", "dispense") at a level relevant to biology, and to provide the nouns like "well A1 of the labware in this well" that allow those verbs to raise to that higher level.
> 
> A big part of the refactors we're engaging in is to split the protocol layer into multiple side-by-side implementations with different levels of input checking, but only really within that layer. We may add more verbs (like a better first-class treatment of "move to") but the protocol layer as a whole retains its job of essentially translating the high-level language of biology into the medium-level language that the hardware controller speaks.
> 
> That means that something like `labware.py` and the rest of the geometry system are, well, more than just utilities. They're in fact fundamental to the split between the layers because they provide the nouns.
> 
> This is maybe a lot of words to say that it doesn't make sense to me to move `geometry` and `module_geometry` to the `protocols` package but keep `labware` in the `protocol_api` area. And it may in fact make the most sense to carve out a different package for the geometry section so that under `opentrons.protocols` we have
> 
> * A subpackage for the nouns of the protocol layer (note: not protocol api)
> * A subpackage for the interfaces to the verbs of the protocol layer
>   
>   * perhaps with further subpackages for implementations, or they're somewhere else
>   * maybe this is where the protocol_api definitions live and we preserve `opentrons.protocol_api` as a dummy with an `__init__` that doesn't do much other than import the lower level packages

This is great input. 

In response to @Laura-Danielle comment i've started splitting the modules apart more. For example, taking `Deck` out of geometry and into its own module. There are a lot of very large files in the `protocol_api` package. Perhaps I'll split up labware as well.

**Comment by sfoster1** [comment]
> Perhaps I'll split up labware as well.
:pray:

**Comment by mcous** [comment]
I like this split as is (or as @Laura-Danielle described), especially as a jumping off point for continued shuffling. I really like how if you ask "what APIs are exposed to the user via the Protocol API?" you can answer with "the APIs in the `protocol_api` directory". This is a good foundation to start from IMO.

My two cents for the rest of the thread below. These thoughts are in the context of continued future work, not necessarily this PR, which is already in a good place for me.

> Seth:
> This is maybe a lot of words to say that it doesn't make sense to me to move geometry and module_geometry to the protocols package but keep labware in the protocol_api area. And it may in fact make the most sense to carve out a different package for the geometry section...

>> Amit:
>> Perhaps I'll split up labware as well.

I think that the "protocol layer" vs "hardware control" layer that @sfoster1 articulated is a good way to approach this, but **for me it makes total sense that `labware.py` (what I would prefer to be called `labware_context.py` for consistency) lives in `protocol_api` because it provides a public interface to the user as part of the protocol API. The explicitly geometry-focused stuff _should_ be elsewhere as part of the non-user facing logic for translating protocol calls to hardware controller calls.

The key for my brain here is that I think **geometry, even in deck coordinates, should not inherently be part of the public protocol API**. Geometry is an implementation detail, that we may sometimes build escape hatches to deal with directly. Those escape hatches should be the exception, though. From that standpoint, the existing `labware.py` contains _a bunch_ of geometry stuff, which in my mind violates the split discussed above and makes me agree with @amitlissack's sentiment that it should be broken up (probably at some point in the future).

**Review by Laura-Danielle — APPROVED** [review]
LGTM

File: api/src/opentrons/protocols/geometry/util.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
+import functools
```
Comment (sfoster1) [inline api/src/opentrons/protocols/geometry/util.py:1]: This is a bit of a nitpick, but if you look at what's actually in this file it's entirely about planning moves. Maybe we could name the file `geometry.planning`?

File: api/src/opentrons/protocols/execution/execute_v3.py [hunk 2]
```diff
@@ -1,8 +1,8 @@
 import logging
 from typing import Dict, TYPE_CHECKING
 
-from .contexts import ProtocolContext, InstrumentContext
-from . import labware
+from opentrons.protocol_api.contexts import ProtocolContext, InstrumentContext
```
Comment (sfoster1) [inline api/src/opentrons/protocols/execution/execute_v3.py:4]: Naming nit - we have a bunch of files named things about execution but they do slightly different things.

`execute.py` takes in a generic protocol which may be either json or python, decides on the appropriate way to run it, and then runs it. It also happens to contain the functions that actually do the running for python protocols. What if we
- Split the things that actually execute python protocols to a new module `execute_python.py`
- Put the rest of `execute.py` in either `__init__` or `execution_control.py` or something?
- Rename `execute_vN` to `execute_json_vN`

Then I think it's a bit clearer from the filenames what they actually do.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
A couple naming nits but overall looks good!

One thing I think we can do (that I couldn't comment inline because it's a pure file rename) is to make `geometry.utils` actually `geometry.planning` - if you look at its contents it's pretty much entirely about planning arc moves and planning thermocycler avoidance moves.

File: api/src/opentrons/protocols/geometry/util.py (hunk 1 above)
Comment (amitlissack) [reply api/src/opentrons/protocols/geometry/util.py:1]: I like it.

**Review by sfoster1 — APPROVED** [review]
Looks great to me! Thanks!

**Review by mcous — APPROVED** [review]

File: api/src/opentrons/protocols/execution/execute_v3.py (hunk 2 above)
Comment (shlokamin) [reply api/src/opentrons/protocols/execution/execute_v3.py]: I like this! I imagine @IanLondon will find this interesting too

**Comment by shlokamin** [comment]
This looks great! My thoughts:

> I really like how if you ask "what APIs are exposed to the user via the Protocol API?" you can answer with "the APIs in the protocol_api directory". This is a good foundation to start from IMO.

Agreeed.

> The key for my brain here is that I think geometry, even in deck coordinates, should not inherently be part of the public protocol API. Geometry is an implementation detail, that we may sometimes build escape hatches to deal with directly. Those escape hatches should be the exception, though

Also agreed — just wanna reiterate that **we do need those escape hatches**. I.e. in PD we need to be able to control robot movements at a more granular level than `aspirate in this location` (like when we delay at a specific z offset from a particular well), and in the future we might need to be able to control the robot's movement at a low level for something else (same logic applies for PAPI). Like @mcous said though:

> Those escape hatches should be the exception

**Review by shlokamin — APPROVED** [review]

**Comment by mcous** [comment]
@shlokamin that's a really good point. I think I should've made a more careful distinction between absolute geometry and relative geometry in that opinion, because something like 5mm from the bottom of the well is 100% a protocol/science concern

**Pull request merged**